### PR TITLE
[Windows] fix thread names in debugger / crash dumps

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

Fixed up / updated the way thread names are set so they can be seen in the debugger.

* The legacy method used by Kodi so far needed a thread id instead of a thread handle (all Windows platforms)

* Added the modern method SetThreadDescription available with Windows 10 1607 and above. The new method adds a few niceties, see https://learn.microsoft.com/en-us/visualstudio/debugger/how-to-set-a-thread-name-in-native-code?view=vs-2022
  * Both method can coexist, but might as well use only the new one when available, with fallback to the legacy method.
  * The new method requires dynamic loading as the function does not exist in older versions of Windows
To avoid impacting performance, LoadLibrary/GetProcAddress is done once and a cached pointer is used for every new thread. Could be simplified if thread creation time is a non-issue.

  * No support added for UWP (covered by legacy method). I'm not familiar with Win RT and didn't find an equivalent that didn't require bundling KernelBase.dll with the app, most likely not a good idea.

* Had to define new Windows builds in order to detect Windows 10 1607 properly

## Motivation and context

Generic thread names in the debugger make debugging more difficult.

## How has this been tested?

Windows 10 22H2 x64 and UWP x64. With old method and new method enabled selectively by special code not included in PR.

## What is the effect on users?

none
## Screenshots (if appropriate):
Before, generic names assigned by Visual Studio:
image

After, Kodi specified names:
image

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/platform/win32/threads/ThreadImplWin.cpp [hunk 1]
```diff
@@ -60,29 +60,34 @@ CThreadImplWin::CThreadImplWin(std::thread::native_handle_type handle) : IThread
...
-
-  __try
-  {
-    RaiseException(MS_VC_EXCEPTION, 0, sizeof(info) / sizeof(ULONG_PTR), (ULONG_PTR *)&info);
-  }
-  __except(EXCEPTION_EXECUTE_HANDLER)
-  {
+    info.dwType = 0x1000;
+    info.szName = name.c_str();
+    info.dwThreadID = GetThreadId(reinterpret_cast<HANDLE>(m_handle));
```
[the remark this answers is not in the document]
Comment (CrystalP) [reply xbmc/platform/win32/threads/ThreadImplWin.cpp]: true, I didn't question the existing code

File: xbmc/platform/win32/threads/ThreadImplWin.cpp [hunk 2]
```diff
@@ -60,29 +60,34 @@ CThreadImplWin::CThreadImplWin(std::thread::native_handle_type handle) : IThread
 
 void CThreadImplWin::SetThreadInfo(const std::string& name)
 {
-  const unsigned int MS_VC_EXCEPTION = 0x406d1388;
-
-#pragma pack(push,8)
-  struct THREADNAME_INFO
+  // Modern way to name threads first (Windows 10 1607 and above)
+  if (!CWIN32Util::SetThreadName(reinterpret_cast<HANDLE>(m_handle), name))
```
Comment (thexai) [inline xbmc/platform/win32/threads/ThreadImplWin.cpp:64]:
```suggestion
  if (!CWIN32Util::SetThreadName(static_cast<HANDLE>(m_handle), name))
```

Same here then (it is preferable use `static_cast` whenever it is sufficient)

Comment (CrystalP) [reply xbmc/platform/win32/threads/ThreadImplWin.cpp]: yes, done

**Review by thexai — APPROVED** [review]
Maybe worth backport fix commit only to Nexus

File: xbmc/platform/win32/threads/ThreadImplWin.cpp [hunk 3]
```diff
@@ -60,29 +60,34 @@ CThreadImplWin::CThreadImplWin(std::thread::native_handle_type handle) : IThread
...
+    // Fallback on legacy method - specially configured exception
+    const unsigned int MS_VC_EXCEPTION = 0x406d1388;
+
+#pragma pack(push, 8)
+    struct THREADNAME_INFO
+    {
+      DWORD dwType; // must be 0x1000
+      LPCSTR szName; // pointer to name (in same addr space)
+      DWORD dwThreadID; // thread ID (-1 caller thread)
+      DWORD dwFlags; // reserved for future use, most be zero
```
Comment (garbear) [inline xbmc/platform/win32/threads/ThreadImplWin.cpp:75]: "m**u**st be zero"

Comment (CrystalP) [reply xbmc/platform/win32/threads/ThreadImplWin.cpp]: yes done

**Comment by CrystalP** [comment]
> Maybe worth backport fix commit only to Nexus
That commit is low risk, why not. I'd have to do a PR against the Nexus branch?

**Comment by thexai** [comment]
Yes a new PR against the Nexus with only commit:
"[Windows] fix thread naming in debugger (needs thread id not thread handle)"

(cherry pick same commit)

And if you want also can do the same with: https://github.com/xbmc/xbmc/pull/23017

**Comment by CrystalP** [comment]
@garbear OK with merge?

**Comment by CrystalP** [comment]
@garbear OK with you to merge?

**Review by garbear — APPROVED** [review]

**Pull request merged**